### ISSUE TITLE
Enable cached test results

### ISSUE THREAD

**Issue by benbrittain** [facebook/buck2#183]
We have a few unit tests that take 60s+ to run, so I was fairly surprised when executing `buck2 test //...` repeatedly invokes those tests even though the inputs haven't been modified. 

Is this working as intended? I know that FB has Tpx internally and I'm not sure if that's the infrastructure which is supposed to handle avoiding reruns or not.

**Comment by stepancheg** [comment]
This is how how buck2 works: it does not cache test results. Maybe it should.

TPX does different things, it can for example skip disabled tests. But it does not cache test execution results either.

**Comment by benbrittain** [comment]
That's too bad, cached test results are a pretty killer feature. Is this logic that belongs in the test runner or as part of Buck2 core? Basically, would it be possible for me to write a test runner that caches results?

**Comment by stepancheg** [comment]
Test runner does not have this information.

I think we should pass it to test runner regardless of who should filter unchanged tests, buck2 or test runner.

But after that, I think it should be `buck2` native option like `buck2 test --skip-unchanged`.

**Comment by alexlian** [comment]
> That's too bad, cached test results are a pretty killer feature. 

This was a pretty contentious discussion internally between a number of teams. Ultimately, given Remote Execution's ability to cache, it was deemed something to at least consider. So, IIRC, we do internally have the option to opt-into caching results.

I do forget where this is implemented, and guess as you suspect that it'd be in the test runner. At best, it's buck2 core passing along the information to RE on behalf of the test runner? Perhaps @krallin might know or point towards someone that does.

**Comment by rbtcollins** [comment]
FWIW, a few folk were considering doing a spike to experiment with Buck2 - we like the focus on developer experience, LSP integrations and the like - but our existing Bazel setup depends quite heavily on not running tests with unchanged inputs in order to deliver our sub-minute builds... 

How much work do you think one would be looking at to get a minimum viable patch for this?

**Comment by ndmitchell** [comment]
Note that you can write your own test rule which runs a command, writes out an artifact (e.g. the logs of the test) and then have your tests run at _build_ time. Then you get everything that works on builds (caching, remote execution etc) for tests. The only thing you lose is the `buck2 build`/`buck2 test` distinction. Perhaps that would be the minimal viable patch, and should be entirely doable in custom Starlark rules.

**Comment by vaskomitanov** [comment]
> This was a pretty contentious discussion internally between a number of teams.

@alexlian I'm curious what are the arguments against caching test results? Is it just because it can be done on RE side or there is something more fundamental?

**Comment by benbrittain** [comment]
> Note that you can write your own test rule which runs a command, writes out an artifact (e.g. the logs of the test) and then have your tests run at _build_ time. 

Even though we're pretty excited about `buck2`, I don't think we're willing to give up the `build/test` distinction and cached tests are pretty important for us.

Is there a somewhat reasonable path to getting something like this supported? If someone was to put together a patch that enables it's support in conjunction with RE would it be accepted even though it's a "contentious" feature? Would a feature like this need to be driven internally?

**Comment by Peter-McLean-Intel** [comment]
I'd like to voice my support for this issue. I come from hardware development land, where unit 'tests' take between 10s of minutes to an hour+. This is a major concern about adopting Buck2, as these long tests should only run if the inputs change. Before a release, every test could be force run, but in typical day-to-day development the time savings are necessary to decrease cycle time.

**Comment by vors** [comment]
I'm very surprised that it's not supported, it's part of RE API and part of what makes systems like buck2 and bazel appealing. Please consider implementing, this is a huge adoption blocker.

**Comment by cjhopman** [comment]
Yeah, I've always been on the side that the right thing to do is to cache test results, I'd love to see us internally move to that model. Maybe I'm biased from having test result caching like 10 years ago with bazel.

I don't think it'd be a particularly extensive change to support this, the model of test integration was done to support tests just being actions like any other (and so should be able to support both RE and caching like any other).

**Comment by dansahlin** [comment]
We are investigating Bazel and Buck2.
It seems that if Buck2 cannot support test caching, we only have one choice.
So, is there any news regarding Buck2 supporting test caching?

**Comment by TroyKomodo** [comment]
This is a pretty major feature which is stopping us from using buck2

**Comment by njaremko** [comment]
I hit this issue too, but refused to use bazel...

It turns out buck2 supports this, so long as you have a custom test runner to manage it, and you wrap the test rule to set the flag. If you use remote execution, and you're project is written in rust, https://github.com/njaremko/quokka will do this for you (there's definitely room for improvement, but it works).
